### PR TITLE
solving problem in auto end amounts calculation

### PR DESCRIPTION
**Pull request by sinahashemy**
because of rounding the depreciation amount in each year 
`Math.round((purchaseAmount / totalDepreciationYears / MONTHS_IN_YEAR) * monthsLeft)`
in last part of depreciation, a state is possible to be remain some integer, on code this situation was predicted 
and handled, but exactly in this segment was better to checked the ratio of 
`depreciationAmount/previousEndAmount >= 0.99`  instead of  `newEndAmount <= 1`, 
like this situation the `newEndAmount` can be greater than `1` and not be true in this conditions 
so return some remain integer instead of `0` in last part.

### CODE REVIEW

File: src/index.ts
Comment (goodfellamilos) [inline src/index.ts:54]: Could you add some tests to verify the fix?

**Comment by sonicdoe** [comment]
Could you provide an example where the old code “fails”? I’ve tried 31 over 3 years (as mentioned in [the comment](https://github.com/kontist/asset-depreciation-calculator/blob/4eb3227a5283c88c08a66ba703d703811423834c/src/index.ts#L52)) and, on the `master` branch, the last year has a `startAmount` of 11 and an `endAmount` of 0, but on your branch the last year has an `endAmount` of 1. Isn’t the former already correct?

File: src/index.ts
Comment (sinahashemy) [reply src/index.ts:54]: Yes, Sure.

Comment (sinahashemy) [reply src/index.ts:54]: I've added 3 test, those if run with previous code, return none zero amount in last part, but in new code, all of them passed successfully.

**Comment by sinahashemy** [comment]
> Could you provide an example where the old code “fails”? I’ve tried 31 over 3 years (as mentioned in [the comment](https://github.com/kontist/asset-depreciation-calculator/blob/4eb3227a5283c88c08a66ba703d703811423834c/src/index.ts#L52)) and, on the `master` branch, the last year has a `startAmount` of 11 and an `endAmount` of 0, but on your branch the last year has an `endAmount` of 1. Isn’t the former already correct?

**Comment by sinahashemy** [comment]
> Could you provide an example where the old code “fails”? I’ve tried 31 over 3 years (as mentioned in [the comment](https://github.com/kontist/asset-depreciation-calculator/blob/4eb3227a5283c88c08a66ba703d703811423834c/src/index.ts#L52)) and, on the `master` branch, the last year has a `startAmount` of 11 and an `endAmount` of 0, but on your branch the last year has an `endAmount` of 1. Isn’t the former already correct?

yes, exactly, I think I should change that more deep, with this change, and only change in this threshold (0.99) it will be ok, but in other case it will be return none zero, I'm working on and solve it.

**Review by sonicdoe — COMMENTED** [review]
Thanks for adding the tests 👍 I now understand that we previously only handled the simple case where the rounding error was small enough.

By the way, I had some trouble following the calculation logic. Therefore, for my own understanding, I rewrote the calculation logic with yet another approach. In case it helps anyone else, here’s my commit: [`2508b28`](https://github.com/kontist/asset-depreciation-calculator/blob/2508b28543dae41085e9c2b0902d3f8b13c4ea18/src/index.ts#L98-L149).

File: src/index.spec.ts
Comment (sinahashemy) [reply src/index.spec.ts:414]: yes, they are same, only have difference in `purchaseAmount` (123000, 1230000).

File: src/index.ts
Comment (sinahashemy) [reply src/index.ts:117]: ok, I've installed it on my VSCode.

File: src/index.ts
Comment (sinahashemy) [reply src/index.ts:100]: Ok,

Comment (sinahashemy) [reply src/index.ts:100]: I got it, and correct the format.

File: src/index.ts
Comment (sinahashemy) [reply src/index.ts:104]: this line can be removed, because in [line 86](https://github.com/kontist/asset-depreciation-calculator/pull/4/files#diff-a2a171449d862fe29692ce031981047d7ab755ae7f84c707aef80701b3ea0c80) `endAmount` initialized with `purchaseAmount`.

File: src/index.ts
Comment (sinahashemy) [reply src/index.ts:42]: Ok, I remove that.

File: src/index.ts
Comment (sinahashemy) [reply src/index.ts]: Because I didn't want this line to effect on first year, I moved this line to end of loop, I agree with you that was added some complexity to code and I also didn't want to remove the `calculate` function, I think this function may be useful in future extensions of code as well as help to better understand of code, so for easier understanding I made some changes and remove this line by calculation the months in each year and storing them on an array.

File: src/index.spec.ts
Comment (sonicdoe) [reply src/index.spec.ts:414]: I’m not sure whether it adds enough additional value then 🤔

File: src/index.ts
Comment (sonicdoe) [inline src/index.ts:105]:
```suggestion
  monthsInEachYear[monthsInEachYear.length - 1] = purchaseMonth > 1 ? purchaseMonth - 1 : MONTHS_IN_YEAR;
```

In [line 99](https://github.com/kontist/asset-depreciation-calculator/pull/4/files#diff-a2a171449d862fe29692ce031981047d7ab755ae7f84c707aef80701b3ea0c80R99), we `purchaseMonth > 1` instead of `purchaseMonth - 1 > 0`. I think it’s easier to understand if we use the same comparison here.

Comment (sinahashemy) [reply src/index.ts:105]: Ok, I'm agree. :-)

**Comment by goodfellamilos** [comment]
@sonicdoe What do we do with your commit? Separate PR?

File: src/index.spec.ts
Comment (sinahashemy) [reply src/index.spec.ts:241]: ok, I change that.

**Review by goodfellamilos — APPROVED** [review]
According to the tests calculation looks correct 👍 Pls wait for @sonicdoe's approval as well.

**Comment by sonicdoe** [comment]
> @sonicdoe What do we do with your commit? Separate PR?

I’m not sure whether it’s a clear win over the current logic, but I’ll open a separate pull request afterwards and we can discuss it there.

File: src/index.spec.ts
Comment (sonicdoe) [reply src/index.spec.ts]: We now have:

- should return endAmount of 0
- should return endAmount of 0 for small amount like 31 - purchase made in Jan
- should return endAmount of 0 for small amount like 31 - purchase made after Jan
- should not return endAmount of -1
- should not return endAmount of 3
- should not return endAmount of 5

I think this is becoming confusing. Perhaps we should focus on the semantics instead of on the pure `endAmount`, for example, “should depreciate rounding errors in the last year” or “should handle remainder”.

In any case, I think it’s fine to stick to the existing convention in this pull request.

File: src/index.spec.ts
Comment (sonicdoe) [inline src/index.spec.ts:340]: Does this test (and the one below with “endAmount of 5”) test anything specific that isn’t already covered by the tests above? In other words, is it just another “example” with different values or does it actually test a specific case?

For example, I understand having tests for `2016-01-01` and `2016-05-01` (because they need to treated quite differently), however, there’s arguably not much more benefit in testing `2016-06-01` or `2016-07-01`.

Comment (sinahashemy) [reply src/index.spec.ts:340]: yes exactly only show the possibility of remaining amount more than 1, like 3 and 5, if in future will be any change in calculation, this is good to be remained this tests to cover that. I agree with you, with this new approach for calculation, this tests may be not much more benefits but may be show the improvement and previous challenges.

**Review by sonicdoe — APPROVED** [review]
Thanks for improving the core logic and fixing these rounding errors 👍

I don’t fully agree with the tests. They test very similar scenarios and descriptions like “should not return endAmount of x” are vague and only make sense in the context of the previous, erroneous calculation.

On a more general note, I wonder whether we really should put all remainders into the last year. For comparison, [Zinsen-berechnen.de](https://www.zinsen-berechnen.de/afa-rechner.php) depreciates remainders evenly across the whole period. In any case, it will never make a big difference and it’s outside of the scope of this pull request.

Finally, before you merge, please apply the two suggestions [here](https://github.com/kontist/asset-depreciation-calculator/pull/4/files#r721170079) and [here](https://github.com/kontist/asset-depreciation-calculator/pull/4/files#r721170209).

**Comment by pstockschlaeder** [comment]
Regarding tests (although it maybe still is a bit over-specific) we could verify that `endAmount` should generally be `0` like so:

```js
  it("endAmount should always be 0", () => {
    const purchaseDates = [
      new Date("2020-01-01"),
      new Date("2020-02-29"),
      new Date("2020-12-31"),
    ];
    purchaseDates.forEach((purchaseDate) =>
      expect(
        calculateDepreciation({
          purchaseAmount: 123000,
          purchaseDate,
          totalDepreciationYears: 7,
        }).slice(-1)[0].endAmount
      ).toEqual(0)
    );
  });
```

With the old erroneous calculation the `endAmount` was changing depending on the `purchaseDate`. We could potentially even feed every day of the year into `purchaseDates`.

**Pull request merged**
